### ISSUE TITLE
MinGit can be incompatible with Git for Windows installation of the same bitness

### ISSUE THREAD

**Issue by luhansen-ms** [git-for-windows/git#4255]
# Summary:
This Issue is in reference to a [ticket](https://developercommunity.visualstudio.com/t/Visual-Studio-2022-embedded-git-installa/10205528) that was filed with us in Visual Studio. However, we believe the underlying issue to be related to mismatched versions of MinGit/Git for Windows. I haven't been able to reproduce the issue myself, but there is a lot of good info provided by the affected user in the linked ticket that might make more sense here. 

## Duplicate info from Linked Ticket
Git for Windows includes a component that needs to be pinned in memory at a known location, so that the Unix-like shell that git bash uses can find it. Anything that moves the component – ASLR, rebasing due to address conflicts, etc. – will cause the component to stop working. The specific failure case here seems to be that when you’ve got multiple versions of Git for Windows installed of the same bitness, if that shared component DLL isn’t binary-identical across those versions, the second one that is loaded into memory will be automatically rebased and stop working.

The affected component is msys-2.0.dll, which is present at (for example):
`C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Git\usr\bin` for a typical Visual Studio installation; and
`C:\Program Files\Git\usr\bin` within a typical Git for Windows installation.

It seems that this problem may have started occurring when VS 2022 began to include 64-bit git components, at version 17.2.0. The problem is heavily dependent on the version of the Git for Windows installation that is side-by-side with VS 2022; with GfW 2.35.2 against VS 2022 17.2.0, the problem is reproducible. Updating VS 2022 to 17.2.7 moves its internal git installation to 2.37.1 and the problem appears to cease; however, it seems likely that further changes to either git installation could cause the problem to recur.

Any git operation taken from within Visual Studio in this state will fail with a base address mismatch in the MSYS component, with this output:
```
0 [main] which (4152) C:\Program Files\Git\usr\bin\which.exe: *** fatal error - cygheap base mismatch detected - 0x18034B408/0x180349408.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
```

## Questions
I haven't found any existing resources mentioning this happening with 64bit git. Does this sound like an expected scenario? Are there any workarounds besides ensuring that all git processes on the system are compatible versions of Git+msys2?
_____

 - [ X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit? 64 bit

```
MinGit: 2.36.0
GfW: 2.35.2
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
Windows 10
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
unknown at this time
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I can ask this in the linked ticket.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other
GfW: unknown
MinGit: VS spawns git.exe process directly via Kernel32.CreateProcess

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
seemingly occurs when GfW and MinGit load conflicting msys2 versions
```

**Comment by rimrul** [comment]
> ## Questions
> I haven't found any existing resources mentioning this happening with 64bit git. Does this sound like an expected scenario?

Yes, this seems to be a likely scenario when using parts of Git for Windows that need a posix-ish environment.

> Are there any workarounds besides ensuring that all git processes on the system are compatible versions of Git+msys2?

The rebasing method described in https://github.com/git-for-windows/git/wiki/32-bit-issues#adjusting-msys-20dlls-address-range-manually should work around this.

Using the busybox based MinGit would theoretically reduce this issue, but it is not currently considered production ready and does not prevent the issue completely IIRC.

The general effort to port more scripted parts of git to builtins should also help to slowly reduce this issue in future versions of Git for Windows.

When we update our Msys2 runtime to Cygwin 3.4 this issue should theoretically cease to exist completely, but that is currently on hold, because it drops support for i686 and has had a bunch of follow-up issues.

> MinGit: VS spawns git.exe process directly via Kernel32.CreateProcess

That's good, that avoids imeadiately loading `msys2-0.dll` for bash. I don't know what commands you run from Visual Studio, but trying to avoid scripted commands would also help. The user could try to reduce this issue by using Git-CMD or Powershell instead of Git-Bash when working with Git for Windows while using MinGit through Visual Studio in parallel.

**Comment by vermiculus** [comment]
Even as most of the address-conflict talk is going over my head, I can add that we're able to reproduce this with `git-fetch` -- a built-in.

It's also worth noting that we have client hooks that invoke git.exe directly (via `Process.Start`) that would be triggered by actions from VS22's MinGit. Would these hooks be considered scripted Git? I would imagine so. In that case, I don't think there's anything the VS team can really do to prevent folks from invoking Git in client hooks -- and using Git in this way certainly has a reasonable expectation of not crashing.

If it's relevant: to get around the fact that VS22 lags substantially behind in Git releases, we also alter PATH early in our hooks such that the system Git is used. I imagine this is not something most folks would need to do.

**Comment by rimrul** [comment]
> It's also worth noting that we have client hooks that invoke git.exe directly (via `Process.Start`) that would be triggered by actions from VS22's MinGit. Would these hooks be considered scripted Git?

It's not what I had in mind when I wrote that part, but it could very well trigger this issue, depending on the shebang.

> I would imagine so. In that case, I don't think there's anything the VS team can really do to prevent folks from invoking Git in client hooks

Right. The Visual studio team has little control over the hooks their users have in their repos. That's why I didn't originally talk about hooks. Invoking git isn't (directly) what's causing this issue. What's causing the issue, but rather invoking conflicting versions of the various unix tools in Git for Windows (which can happen through git, like for example with hooks). Basically anything in `/usr/` will conflict with the same tools in other copies of Git for Windows.

> and using Git in this way certainly has a reasonable expectation of not crashing.

Sure. We just don't have a reasonable way to make it not crash at the moment.

> If it's relevant: to get around the fact that VS22 lags substantially behind in Git releases, we also alter PATH early in our hooks such that the system Git is used. I imagine this is not something most folks would need to do.

That might exacerbate the problem in your environment. Assuming your hooks are written as shell scripts their shebang is probably `#!/bin/bash` or similar, which will load the Msys2 runtime of MinGit. You'll then call the system Git with its unix tools that will in turn load a conflicting Msys2 runtime.

**Comment by dscho** [comment]
I investigated this a bit.

The root cause of these issues is that the MSYS2 runtime (or for that matter, the Cygwin runtime, from which the MSYS2 runtime is forked) needs to emulate certain POSIX functionality, e.g. the `fork()` syscall, and said emulation requires the child processes to talk to their parent processes "before the child process starts up for real". This communication happens via shared memory that is called the "Cygwin heap". And naturally, such communication will only succeed if both parent and child process have a very, very similar idea of the shape of that Cygwin heap. If they have even slightly different ideas, some of those bytes in that shared memory are misinterpreted by one side and chaos ensues. Or crashes. Or aborts with a cryptic message about some "cygheap base mismatch, whatever that is".

Now, the Cygwin runtime is very careful to version the definition of that Cygwin heap so that even different Cygwin runtimes have a chance to talk to each other. It is quite possible that by rebasing the MSYS2 runtime's custom changes on top of newer Cygwin runtime versions, we miss some steps and inadvertently break that shared memory format versioning.

I implemented a pretty big hammer [here](https://github.com/dscho/msys2-runtime/commit/75cd57a982724352a9c13e439d67b4fed1d407df), which seems to do the job by telling _any_ two different MSYS2 runtime versions that they simply cannot talk to each other.

This would most likely address this here ticket, but it might be too big of a hammer (and therefore have unfortunate consequences that I simply haven't had a chance to think through yet).

I will let this simmer for a few days, welcoming any insights anybody else might have, and then try to finalize a fix.

**Comment by luhansen-ms** [comment]
Thanks all for the contributions to this issue, and to @dscho for getting that patch in. We'll look forward to using it!

**Comment by dscho** [comment]
@luhansen-ms you're welcome!

Could I ask for a favor in return? It would be good to get some confirmation that installing [the latest snapshot](https://wingit.blob.core.windows.net/files/index.html) does indeed fix the problem. Would you mind trying that and reporting back the result?

**Comment by luhansen-ms** [comment]
@dscho Indeed, I haven't been able to reproduce the issue myself, but we'll be syncing tomorrow with our user who can reproduce this consistently. We'll try to validate this with them over the next few days.

This was also before the topic of scripted (loading bash/msys) vs built-ins came up in this thread, so I'll give it another go myself.

**Comment by filefull** [comment]
@dscho, testing with the latest snapshot I am unable to reproduce the bug, and haven't seen any other side effects or odd behaviors. Based on my understanding of the root cause and what you've done here, this looks like it should resolve things for us - thank you.

(I should however note that I am currently unable to reproduce the bug with _released_ builds, either... but that's not particularly unusual - in my experience this has been pretty finicky. I've seen pairings of VS 2022 and GfW that will _never_ have trouble because their git versions are synced, and pairings that will _usually_ have trouble because they aren't, but none that will _always_ have trouble.)

**Comment by luhansen-ms** [comment]
Thanks for testing this out @filefull. 

@dscho Would you expect that only one of the "mismatched versions" would need to be updated to take advantage of the fix? Said another way, either the first or second process to load msys2 could be the version with the fix, since the patched version would have generated the new magic number, and therefore either version would bail out before ever attempting to treat the "first" process as valid msys2/cygwin process to share the heap with?

**Comment by dscho** [comment]
> Would you expect that only one of the "mismatched versions" would need to be updated to take advantage of the fix?

@luhansen-ms yes.

**Comment by luhansen-ms** [comment]
@dscho Is this looking like it would make its way into 2.40?

**Comment by dscho** [comment]
@luhansen-ms it already made it into v2.39.2...
